pkg/pillar/localcommand/agent.go

```diff
@@ -0,0 +1,564 @@
+// LocalCmdAgent tasks that read from this map do not need to take the lock,
+// since they are paused while LPS address discovery is in progress.
+type lpsAddresses struct {
+	addrsByIface    map[string][]lpsAddress // key = interface name, value = addresses
+	upToDate        bool
+	addrDiscoveryMx sync.Mutex
+}
+
+// Returns true if there are no stored LPS addresses.
+func (la *lpsAddresses) empty() bool {
+	for _, addrs := range la.addrsByIface {
+		if len(addrs) > 0 {
+			return false
+		}
+	}
+	return true
+}
+
+// taskTicker wraps a flextimer.FlexTickerHandle with synchronization
+// and a throttle flag. It allows dynamically adjusting ticker intervals,
+// forcing immediate ticks, and temporarily suppressing ticks when throttled.
+type taskTicker struct {
+	sync.RWMutex
+	tickerHandle flextimer.FlexTickerHandle
+	throttled    bool
+}
+
+// newTaskTicker creates a new taskTicker with a randomized firing interval.
+// The actual interval is chosen randomly between 0.8×1.1×interval (0.88×interval)
+// and 1.1×interval, which introduces jitter to avoid synchronized task execution
+// across multiple Go routines.
+func newTaskTicker(interval time.Duration) *taskTicker {
+	maxTime := 1.1 * float64(interval)
+	minTime := 0.8 * maxTime
+	t := &taskTicker{
+		tickerHandle: flextimer.NewRangeTicker(
+			time.Duration(minTime), time.Duration(maxTime)),
+	}
+	return t
+}
+
+// update sets the throttle state and recalculates the ticker's randomized interval.
+// When throttled is true, callers are expected to pass a longer interval (e.g. 1h
+// instead of 1m), and the ticker will continue to fire within the new randomized
+// [0.8×1.1×interval, 1.1×interval] range.
+func (t *taskTicker) update(throttle bool, interval time.Duration) {
+	t.Lock()
+	defer t.Unlock()
+	t.throttled = throttle
+	maxTime := 1.1 * float64(interval)
+	minTime := 0.8 * maxTime
+	t.tickerHandle.UpdateRangeTicker(time.Duration(minTime), time.Duration(maxTime))
+}
+
+// tickerIsThrottled reports whether the ticker is currently throttled.
+func (t *taskTicker) tickerIsThrottled() bool {
+	t.RLock()
+	defer t.RUnlock()
+	return t.throttled
+}
+
+// tickNow forces the ticker to fire immediately, unless throttled.
+func (t *taskTicker) tickNow() {
+	if t.tickerIsThrottled() {
+		return
+	}
+	t.tickerHandle.TickNow()
+}
+
+// tickerChan returns the ticker’s channel for receiving tick events.
+func (t *taskTicker) tickerChan() <-chan time.Time {
+	return t.tickerHandle.C
+}
+
+// taskControl provides thread-safe pause/resume functionality for concurrent tasks.
+// It allows multiple callers to pause task execution and requires all of them
+// to resume before tasks can continue running. Tasks can run in parallel when
+// not paused.
+type taskControl struct {
+	beforeStart  func()
+	pauseCounter int
+	pauseMx      sync.Mutex
+	taskMx       sync.RWMutex
+}
+
+// pause increments the pause counter and blocks new tasks from starting.
```

Comment:
Do we care about any tasks which might already be running?

Reply:
`pause` blocks until the already running tasks complete and release the lock.
(added this to the comment)